core/src/main/java/org/elasticsearch/index/IndexModule.java

```diff
@@ -360,6 +361,16 @@ public IndexService newIndexService(NodeEnvironment environment, IndexService.Sh
     }
 
     /**
+     * creates a new mapper service to do administrative work like mapping updates. This *should not* be used for document parsing.
+     * doing so will result in an exception.
+     */
+    public MapperService newIndexMapperService(MapperRegistry mapperRegistry) throws IOException {
+        return new MapperService(indexSettings, analysisRegistry.build(indexSettings),
+            new SimilarityService(indexSettings, similarities), mapperRegistry,
+            () -> { throw new UnsupportedOperationException("no index query shard context available"); });
```

Comment:
I wonder if we should throw and AssertionError here instead... it should not happen and should not be caught?